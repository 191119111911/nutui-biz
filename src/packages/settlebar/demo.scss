.demo {
    [class*="safe-area-bottom"] {
        position: relative;
    }
}