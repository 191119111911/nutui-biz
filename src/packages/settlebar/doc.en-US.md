#  SettleBar 结算栏

### 介绍

常见于购物车页底部，包括全选、合计与「去结算」按钮。也适用于结算页底部的提交订单栏。

### 安装

```javascript
import { SettleBar } from '@nutui/nutui-biz';
```

## 代码演示

### 基本用法

:::demo

```ts
import  React from 'react';
import { SettleBar } from '@nutui/nutui-biz';

const App = () => {
  return (
    <SettleBar />
  );
};
export default App;
```

:::

### 对齐方式

:::demo

```ts
import  React from 'react';
import { SettleBar } from '@nutui/nutui-biz';

const App = () => {
  return (
    <SettleBar totalAlign="left" />
  );
};
export default App;
```

:::

### 禁用状态

:::demo

```ts
import  React from 'react';
import { SettleBar } from '@nutui/nutui-biz';

const App = () => {
  return (
    <SettleBar disabled />
  );
};
export default App;
```

:::

### 加载状态

:::demo

```ts
import  React from 'react';
import { SettleBar } from '@nutui/nutui-biz';

const App = () => {
  return (
    <SettleBar loading />
  );
};
export default App;
```

:::

### 提交订单

:::demo

```ts
import  React from 'react';
import { SettleBar } from '@nutui/nutui-biz';

const App = () => {
  return (
    <SettleBar customSelectAll="" showZero={false} totalText="总计" settleButtonText="提交订单" />
  );
};
export default App;
```

:::

### 去结算数量和单位

:::demo

```ts
import  React from 'react';
import { SettleBar } from '@nutui/nutui-biz';

const App = () => {
  return (
    <SettleBar settleCount="100" settleUnit="个" />
  );
};
export default App;
```

:::

### 自定义合计额外区域内容

:::demo

```ts
import  React from 'react';
import { SettleBar } from '@nutui/nutui-biz';

const App = () => {
  return (
    <SettleBar customTotalExtra={<div style={{fontSize: '12px'}}>已减 ¥30.00</div>} />
  );
};
export default App;
```

:::

### 带有警告信息

:::demo

```ts
import  React from 'react';
import { SettleBar } from '@nutui/nutui-biz';

const App = () => {
  const customWarningHtml = () => {
    return <div style={{display: 'flex', height: '100%', alignItems: 'center', fontSize: '12px', justifyContent: 'center', color: 'red'}}>此商品无货！</div>
  }
  
  return (
    <SettleBar customWarning={customWarningHtml()} />
  );
};
export default App;
```

:::


## API

### Props


| 字段    | 说明                                       | 类型    | 默认值    |
|---------|--------------------------------------------|---------|-----------|
| total   | 合计价格                                 | Number  | 0         |
| totalText     | 合计文案                   | String  | `合计`    |
| totalAlign | 合计区域对齐方式，可选值：`left`、`right`                       | String  | `right`      |
| settleCount     | 结算数量                               | Number | 0    |
| settleUnit  | 结算单位                                  | String | -    |
| settleButtonText     | 结算按钮文案 | String  | `去结算`     |
| disabled   | 结算按钮是否置灰| Boolean  | `true`      |
| loading   | 结算按钮是否加载中| Boolean  | `false`      |
| showZero   | 数量为0时是否展示                                 | Boolean  | `true`          |
| safeAreaInsetBottom   | 是否开启底部安全区适配                                 | Boolean  | `true`          |
| placeholder   | 是否在标签位置生成一个等高的占位元素                                 | Boolean  | `false`          |
| customTotal | 合计区域自定义 | React.ReactNode  | -          |
| customWarning | 上面提示内容自定义 | React.ReactNode  | -          |
| customSelectAll | 全选内容自定义 | React.ReactNode  | -          |
| customTotalPrice | 合计价格内容自定义 | React.ReactNode  | -          |
| customTotalExtra | 合计额外区域自定义 | React.ReactNode  | -          |
| customButton | 按钮内容自定义 | React.ReactNode  | -          |


## Events
| 字段 | 说明 | 回调参数 |
|----- | ----- | -----  |
| onSelectAll | 全选按钮点击事件 |  event: Event |
| onSubmit | 去结算按钮点击事件 |  event: Event |
| onDelete | 删除事件，isEdit 为 true 时生效 |  event: Event |