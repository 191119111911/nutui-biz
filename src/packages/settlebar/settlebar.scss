@import '../../styles/base.scss';
@import '../../styles/mixins/text-ellipsis.scss';

.nb-settle-bar {
    position: fixed;
    bottom: 0;
    left: 0;
    right: 0;

    &__warning {
        position: relative;
        height: 38px;

        &-mask,
        &-content {
            position: absolute;
            top: 0;
            left: 0;
            height: 100%;
            width: 100%;
        }

        &-mask {
            z-index: 1;
            background-color: #fffbe8;
            opacity: .8;
        }

        &-content {
            z-index: 2;
        }
    }

    &__main {
        padding: 6px 12px;
        display: flex;
        align-items: center;
        justify-content: space-between;

        .nut-checkbox {
            cursor: pointer;

            &__label {
                margin-left: 10px;
            }
        }

        &-select-all {
            margin-right: 12px;
            flex-shrink: 0;
        }
    
        &-total {
            flex: 1;
            display: flex;
            flex-direction: column;
            margin-right: 12px;

            &-inner {
                max-width: 120px;
                @include text-ellipsis();
            }
        }

        .btn {
            border: 1px solid #8c8c8c;
            border-radius: 15px;
            font-size: 12px;
            line-height: 30px;
            padding: 0 19px;
        }
    
        &-buy {
            display: flex;
            justify-content: center;
            align-items: center;
            background-color: #f2270c;
            background-image: linear-gradient(135deg,#f2140c,#f2270c 70%,#f24d0c);
            border-radius: 20px;
            color: #fff;
            font-size: 14px;
            font-weight: 700;
            height: 38px;
            width: 113px;
            cursor: pointer;
            flex-shrink: 0;
    
            &.disabled {
                opacity: .3;
                cursor: not-allowed;
            }
        }
    
        &-num {
            font-weight: 400;
        }
    }
}