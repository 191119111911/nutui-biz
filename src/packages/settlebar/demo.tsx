import React from 'react'
import { SettleBar } from './settlebar'
import { useTranslate } from '../../sites/assets/locale'

interface T {
}
const SettleBarDemo = () => {
  const customWarningHtml = () => {
    return <div style={{display: 'flex', height: '100%', alignItems: 'center', fontSize: '12px', justifyContent: 'center', color: 'red'}}>此商品无货！</div>
  }

  return (
    <>
      <div className="demo">
        <h2>基本用法</h2>
        <SettleBar />
        <h2>对齐方式</h2>
        <SettleBar totalAlign="left" />
        <h2>禁用状态</h2>
        <SettleBar disabled />
        <h2>加载状态</h2>
        <SettleBar loading />
        <h2>提交订单</h2>
        <SettleBar customSelectAll="" noCount={true} totalText="总计" settleButtonText="提交订单" />
        <h2>去结算数量和单位</h2>
        <SettleBar settleCount="100" settleUnit="个" />
        <h2>自定义合计额外区域内容</h2>
        <SettleBar customTotalExtra={<div style={{fontSize: '12px'}}>已减 ¥30.00</div>} />
        <h2>带有警告信息</h2>
        <SettleBar customWarning={customWarningHtml()} />
      </div>
    </>
  )
}

export default SettleBarDemo
