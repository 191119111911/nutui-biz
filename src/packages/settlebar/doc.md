#  SettleBar 结算栏

### 介绍

常见于购物车页底部，包括全选、合计与「去结算」按钮。也适用于结算页底部的提交订单栏。

### 安装

```javascript
import { SettleBar } from '@nutui/nutui-biz';
```

## 代码演示

### 基本用法

:::demo

```ts
import  React from 'react';
import { SettleBar } from '@nutui/nutui-biz';
import { Toast } from '@nutui/nutui-react';
import "@nutui/nutui-biz/dist/styles/demo.css";

const App = () => {
  return (
    <div className="demo">
      <SettleBar
        total={100}
        onClickButton ={() => Toast.text('点击按钮')}
      />
    </div>
  );
};
export default App;
```

:::

### 对齐方式

:::demo

```ts
import  React from 'react';
import { SettleBar } from '@nutui/nutui-biz';
import { Toast } from '@nutui/nutui-react';
import "@nutui/nutui-biz/dist/styles/demo.css";

const App = () => {
  return (
    <div className="demo">
      <SettleBar 
        total={100}
        totalAlign="left"
        onClickButton ={() => Toast.text('点击按钮')}
      />
    </div>
  );
};
export default App;

```

:::

### 禁用状态

:::demo

```ts
import  React from 'react';
import { SettleBar } from '@nutui/nutui-biz';
import "@nutui/nutui-biz/dist/styles/demo.css";

const App = () => {
  return (
    <div className="demo">
      <SettleBar 
        total={100}
        disabled
      />
    </div>
  );
};
export default App;

```

:::

### 加载状态

:::demo

```ts
import  React from 'react';
import { SettleBar } from '@nutui/nutui-biz';
import "@nutui/nutui-biz/dist/styles/demo.css";

const App = () => {
  return (
    <div className="demo">
      <SettleBar 
        total={100}
        loading 
      />
    </div>
  );
};
export default App;

```

:::

### 提交订单

:::demo

```ts
import  React from 'react';
import { SettleBar } from '@nutui/nutui-biz';
import { Toast } from '@nutui/nutui-react';
import "@nutui/nutui-biz/dist/styles/demo.css";

const App = () => {
  return (
    <div className="demo">
      <SettleBar 
        total={100}
        customSelectAll="" 
        showZero={false} 
        totalText="总计" 
        settleButtonText="提交订单" 
        onClickButton ={() => Toast.text('点击按钮')}
      />
    </div>
  );
};
export default App;

```

:::

### 去结算数量

:::demo

```ts
import  React from 'react';
import { SettleBar } from '@nutui/nutui-biz';
import { Toast } from '@nutui/nutui-react';
import "@nutui/nutui-biz/dist/styles/demo.css";

const App = () => {
  return (
    <div className="demo">
      <SettleBar 
        total={100}
        settleCount="100" 
        onClickButton ={() => Toast.text('点击按钮')}
      />
    </div>
  );
};
export default App;

```

:::

### 自定义合计额外区域内容

:::demo

```ts
import  React from 'react';
import { SettleBar } from '@nutui/nutui-biz';
import { Toast } from '@nutui/nutui-react';
import "@nutui/nutui-biz/dist/styles/demo.css";

const App = () => {
  return (
    <div className="demo">
      <SettleBar 
        total={100}
        customTotalExtra={<div style={{fontSize: '12px'}}>已减 ¥30.00</div>} 
        onClickButton ={() => Toast.text('点击按钮')}
      />
    </div>
  );
};
export default App;

```

:::

### 带有警告信息

:::demo

```ts
import  React from 'react';
import { SettleBar } from '@nutui/nutui-biz';
import { Toast } from '@nutui/nutui-react';
import "@nutui/nutui-biz/dist/styles/demo.css";

const App = () => {
  const customWarningHtml = () => {
    return <div style={{display: 'flex', height: '100%', alignItems: 'center', fontSize: '12px', justifyContent: 'center', color: 'red'}}>此商品无货！</div>
  }
  
  return (
    <div className="demo">
      <SettleBar 
        total={100}
        customWarning={customWarningHtml()} 
        onClickButton ={() => Toast.text('点击按钮')}
      />
    </div>
  );
};
export default App;

```

:::


## API

### Props


| 字段    | 说明                                       | 类型    | 默认值    |
|---------|--------------------------------------------|---------|-----------|
| total   | 合计价格                                 | number \| string | 0         |
| totalText     | 合计文案                   | string  | `合计`    |
| totalAlign | 合计区域对齐方式，可选值：`left`、`right`                       | string  | `right`      |
| settleCount     | 结算数量                               | number \| string | 0    |
| settleButtonText     | 结算按钮文案 | string  | `去结算`     |
| disabled   | 结算按钮是否置灰| boolean  | `false`      |
| loading   | 结算按钮是否加载中| boolean  | `false`      |
| showZero   | SettleCount 为 0 时，是否还展示结算按钮中的数量                                 | boolean  | `true`          |
| safeAreaInsetBottom   | 是否开启iphone系列全面屏底部安全区适配                                 | boolean  | `true`          |
| placeholder   | 固定在底部时，是否在标签位置生成一个等高的占位元素                                 | boolean  | `false`          |
| customTotal | 合计区域自定义 | ReactNode  | -          |
| customWarning | 上面提示内容自定义 | ReactNode  | -          |
| customSelectAll | 全选内容自定义 | ReactNode  | -          |
| customTotalPrice | 合计价格内容自定义 | ReactNode  | -          |
| customTotalExtra | 合计额外区域自定义 | ReactNode  | -          |
| customButton | 按钮内容自定义 | ReactNode  | -          |
| isCheckedAll | 全选按钮是否选中 | boolean  | `false`      |


### Events
| 字段 | 说明 | 回调参数 |
|----- | ----- | -----  |
| onSelectAll | 全选按钮点击事件 |  checked |
| onClickButton | 按钮点击事件 |  - |