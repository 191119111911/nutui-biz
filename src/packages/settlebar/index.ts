import { SettleBar } from './settlebar'

export default SettleBar
