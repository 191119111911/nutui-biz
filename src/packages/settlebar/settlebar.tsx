import React, {
  FunctionComponent,
  CSSProperties,
  useRef,
  useState,
  useEffect,
  ReactNode
} from 'react'
import { useConfig } from '@/packages/configprovider'
import classNames from 'classnames';
import { Icon, Checkbox, IconProps, CheckboxProps } from '@nutui/nutui-react';
import {cn2} from '@/utils/bem'
import {getRect} from '@/utils/useClientRect'

import { IComponent } from '@/utils/typings'

export interface SettleBarProps extends IComponent {
  total: number
  totalText: string
  totalAlign: string
  settleButtonText: string
  disabled: boolean
  loading: boolean
  className: string
  style: CSSProperties
  safeAreaInsetBottom: boolean
  placeholder: boolean
  settleCount: number | string
  settleUnit: string
  showZero: boolean
  noCount: boolean
  customWarning: ReactNode
  customTotal: ReactNode
  customSelectAll: ReactNode
  customTotalPrice: ReactNode
  customTotalExtra: ReactNode
  customButton: ReactNode
  iconProps: Partial<IconProps>
  checkboxProps: CheckboxProps
  onSettle: () => void
  onSelectAll: (checked: boolean) => void
}

export const SettleBar: FunctionComponent<
  Partial<SettleBarProps>
> = (props) => {
  const { locale } = useConfig()
  const root = useRef(null)
  const {
    total = 0,
    totalText = locale.settleBar.totalText,
    totalAlign = 'right',
    settleButtonText = locale.settleBar.settleButtonText,
    disabled = false,
    loading = false,
    className,
    style,
    settleCount = 0,
    settleUnit = '',
    customWarning,
    safeAreaInsetBottom = true,
    placeholder = false,
    showZero = true,
    noCount = false,
    customTotal,
    customSelectAll,
    customTotalPrice,
    customButton,
    iconProps,
    checkboxProps,
    customTotalExtra = '',
    onSettle = () => {},
    onSelectAll = (checked: boolean) => {},
    ...rest
  } = {
    ...props,
  }

  const handleSettle = () => {
    !disabled && !loading && onSettle && onSettle()
  }

  let totalStyle = {
    alignItems: totalAlign === 'left' ? 'flex-start' : 'flex-end',
    textAlign: totalAlign
  } as CSSProperties

  const handleSelectAll = (checked: boolean) => {
    onSelectAll && onSelectAll(checked)
  }

  const b = cn2('settle-bar')

  const [height,setHeight] = useState(0)

  useEffect(() => {
    if(root.current) {
      setHeight(getRect(root.current).height)
    }
  }, ['height'])

  const renderCountAndUnit = () => {
    if(showZero || settleCount !== 0) {
      return <span className={b('main-num')}>({settleCount}{settleUnit})</span>
    }
  }

  const renderButton = () => {
    return customButton || <div 
      className={classNames(`${b('main-buy')} ${(disabled || loading) ? 'disabled' : ''}`)}
      onClick={handleSettle}
    >
      {
        loading ? <Icon name='loading' {...iconProps} /> :  <>{settleButtonText}{!noCount && renderCountAndUnit()}</>
      }
    </div>
  }

  const renderSelectAll = () => {
    if(!customSelectAll && customSelectAll !== undefined) return null;

    return <div className={b('main-select-all')}>{customSelectAll ? customSelectAll : <Checkbox label={locale.settleBar.selectAll} onChange={handleSelectAll} {...checkboxProps} />}</div>
  }

  const renderSettleBar = () => {
    return <div ref={root} className={classNames([b(),className,{'nut-biz-safe-area-bottom':safeAreaInsetBottom}])} style={style} {...rest}>
      {
        customWarning && 
        <div className={b('warning')}>
          <div className={b('warning-mask')}></div>
          <div className={b('warning-content')}>{customWarning}</div>
        </div>
      }
      <div className={b('main')}>
        {renderSelectAll()}
        <div className={b('main-total')} style={totalStyle}>
          {
            customTotal ? customTotal : <>
              {
                customTotalPrice ? customTotalPrice : <div>
                  <span>{totalText}：</span>
                  <span>¥{total}</span>
                </div>
              }
              {customTotalExtra}
            </>
          }
        </div>
        {renderButton()}
      </div>
    </div>
  }

  const renderSettleBarWithPlaceholder = () => {
    return <div style={{height}} className={`${b('')}--placeholder`}>
      {renderSettleBar()}
    </div>
  }

  return (
    placeholder ? renderSettleBarWithPlaceholder() : renderSettleBar()
  )
}

SettleBar.displayName = 'NutSettleBar'
