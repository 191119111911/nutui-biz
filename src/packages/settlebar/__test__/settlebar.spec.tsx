import React from 'react'
import { render, fireEvent } from '@testing-library/react'
import '@testing-library/jest-dom'
import { SettleBar } from '../settlebar'

test('test base', async () => {
  const { container } = render(
    <SettleBar />
  )
  const settleBar = container.querySelector('.nb-settle-bar')
  
  expect(settleBar).toHaveClass('nut-biz-safe-area-bottom')
})

test('test totalAlign', async () => {
  const { container } = render(
    <SettleBar totalAlign="left" />
  )
  const total = container.querySelector('.total')
  
  expect(total).toHaveAttribute('style', 'align-items: flex-start;')
})

test('test disable', async () => {
  const { container } = render(
    <SettleBar disabled />
  )
  const total = container.querySelector('.buy')
  expect(total).toHaveClass('disabled')
})

test('test loading', async () => {
  const { container } = render(
    <SettleBar loading />
  )
  const total = container.querySelector('.buy')
  const loadingIcon = container.querySelector('.nut-icon-loading') as HTMLElement
  expect(total).toHaveClass('disabled')
  expect(loadingIcon).toBeTruthy()
})

test('test totalText and settleButtonText', async () => {
  const { container } = render(
    <SettleBar totalText="总计" settleButtonText="jiesuan" />
  )
  const settleBarMainData = container.querySelector('.nb-settle-bar__main')
  expect(settleBarMainData).toHaveTextContent('总计')
  expect(settleBarMainData).toHaveTextContent('jiesuan')
})

test('test settleCount and settleUnit', async () => {
  const { container } = render(
    <SettleBar settleCount="100" settleUnit="个" />
  )
  const settleBarMainData = container.querySelector('.num')
  expect(settleBarMainData).toHaveTextContent('100个')
})

test('test customTotal', async () => {
  const { container } = render(
    <SettleBar customTotal={<div>自定义合计</div>} />
  )
  const settleBarMainData = container.querySelector('.total')
  expect(settleBarMainData).toHaveTextContent('自定义合计')
})

test('test placeholder', async () => {
  const { container } = render(
    <SettleBar placeholder />
  )

  expect(container.querySelector('.nut-settle-bar--placeholder')?.innerHTML).toMatchSnapshot()
})

test('test customWarning', async () => {
  const customWarningHtml = () => {
    return <div style={{display: 'flex', height: '100%', alignItems: 'center', justifyContent: 'center', color: 'red'}}>此商品无货！</div>
  }

  const { container } = render(
    <SettleBar customWarning={customWarningHtml()} />
  )
  const settleBarWarning = container.querySelector('.nb-settle-bar__warning')
  expect(settleBarWarning).toBeInTheDocument()
})