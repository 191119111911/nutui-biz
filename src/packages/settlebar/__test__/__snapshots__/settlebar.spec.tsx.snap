// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test placeholder 1`] = `"<div class=\\"nut-settle-bar nut-biz-safe-area-bottom\\"><div class=\\"nut-settle-bar-main\\"><div class=\\"select-all\\"><div class=\\"nut-checkbox \\"><i class=\\"nutui-iconfont nut-icon nut-icon-check-normal nut-checkbox__icon--unchecked\\" style=\\"font-size: 18px; height: 18px; width: 18px;\\"></i><span class=\\"nut-checkbox__label \\">全选</span></div></div><div class=\\"total\\" style=\\"align-items: flex-end; text-align: right;\\"><div class=\\"total-main\\"><span>合计：</span><span>¥0</span></div></div><div class=\\"buy \\">去结算<span class=\\"num\\">(0)</span></div></div></div>"`;
